presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java

```diff
@@ -1136,6 +1136,37 @@ public void testArrayConstructor()
                 "array_constructor((bound_long + 0), (unbound_long + 1), NULL)");
     }
 
+    @Test
+    public void testRowConstructor()
+    {
+        optimize("ROW(NULL)");
+        optimize("ROW(1)");
+        optimize("ROW(unbound_long + 0)");
+        optimize("ROW(unbound_long + unbound_long2, unbound_string, unbound_double)");
+        optimize("ROW(unbound_boolean, FALSE, ARRAY[unbound_long, unbound_long2], unbound_null_string, unbound_interval)");
+        optimize("ARRAY [ROW(unbound_string, unbound_double), ROW(unbound_string, 0.0)]");
+        optimize("ARRAY [ROW('string', unbound_double), ROW('string', bound_double)]");
+        optimize("ROW(ROW(NULL), ROW(ROW(ROW(ROW('rowception')))))");
+        optimize("ROW(unbound_string, bound_string)");
+
+        // Auto casting does not fully work in the following cases, but can be solved with explicit cast
+
+        // these two tests don't work because one type is varchar(17) and the other varchar
+        // optimize("ARRAY [ROW(unbound_string, unbound_double), ROW(bound_string, 0.0)]");
+        // optimize("ARRAY [ROW(bound_string, 0.0), ROW(unbound_string, unbound_double)]");
+        optimize("ARRAY [ROW(unbound_string, unbound_double), ROW(CAST(bound_string AS VARCHAR), 0.0)]");
+        optimize("ARRAY [ROW(CAST(bound_string AS VARCHAR), 0.0), ROW(unbound_string, unbound_double)]");
+
+        // these two tests don't work because the CAST from NULL generate a row with uppercase
+        // field names: row(FIELD0 varchar, FIELD1 double), while the row expression
+        // generate a with with lowercase field names (row(field0 varchar, field1 double))
+        // Notice that parseRowTypeSignature calls createNamedTypeParameters in TypeSignature
+        // optimize("ARRAY [NULL, ROW(unbound_string, unbound_double)]");
+        // optimize("ARRAY [ROW(unbound_string, unbound_double), NULL]");
```

Comment:
Ya, generally I don't like having commented out tests because we tend to forget about them. With an issue we're more likely to remember, and can track it